### PR TITLE
Visible scroll-top button and smooth scroll on privacy-policy page

### LINKED ISSUE #106: Make up-arrow visible on privacy-policy page and smooth scroll
### Observered Behaviour
The up-arrow is not visible on privacy-page when scrolled at the bottom of the page.

### Expected Behaviour 
The up-arrow should be visible and a smooth scroll effect to it.

@RishabhJain2018

### LINKED ISSUE #107: Add scroll bar for navigating through questions on privacy-policy page
### Observed Behaviour
There is a scroll bar for navigating through questions on Privacy-Policy page which is not functional.

### Expected Behaviour
Add a functional scroll bar for navigating through questions.

@RishabhJain2018

### PR DESCRIPTION
**Pull request by sanketbansal**
Fix #106 
Fix #107

Link to live demo: https://pr-111-evalai.surge.sh

### For issue 106:
- Added Z index value to the scroll top button.
- Reduce the z-index valu of the footer to 0.
- Implemented smooth scroll for scroll top button.
- Made the main content scroll smoothly on clicking nav content.

 

Screenshot (40)

 
### For issue 107
- Added the scroll bar in the nav container of privacy policy page.
- CSS property max-height has been added in the nav container div.

Screenshot (44)

### CODE REVIEW

File: src/app/components/nav/footer/footer.component.scss [hunk 1]
```diff
@@ -11,7 +11,7 @@
   overflow-x: hidden;
   background-repeat: no-repeat;
   background-size: 100% 100%;
-  z-index: 10;
+  z-index: 0;
```
Comment (gautamjajoo) [inline src/app/components/nav/footer/footer.component.scss:14]: Why are you decreasing footer's `z-index`, instead increase for the button.

File: src/app/components/privacy-policy/privacy-policy.component.scss [hunk 2]
```diff
@@ -50,10 +50,11 @@
     }
     .scroll-to-top {
       position: fixed;
-      top: 50%;
-      right: 4%;
+      bottom: 2%;
```
Comment (gautamjajoo) [inline src/app/components/privacy-policy/privacy-policy.component.scss]: Why was this change made?

**Review by gautamjajoo — COMMENTED** [review]
Remove the changes other than solving the issue, a PR is already up for it!

File: src/app/components/nav/footer/footer.component.scss (hunk 1 above)
Comment (sanketbansal) [reply src/app/components/nav/footer/footer.component.scss:14]: Because scroll button is child of  "<div class="privacy-content>" and <app-footer> is sibling of that div.

Comment (sanketbansal) [reply src/app/components/nav/footer/footer.component.scss:14]: `<app-footer>`

File: src/app/components/privacy-policy/privacy-policy.component.scss (hunk 2 above)
Comment (sanketbansal) [reply src/app/components/privacy-policy/privacy-policy.component.scss]: It positions the scroll top button at the bottom right of the privacy page

Comment (gautamjajoo) [reply src/app/components/privacy-policy/privacy-policy.component.scss]: Maybe the earlier one is better! You could revert this change.

**Comment by gautamjajoo** [comment]
@sanketbansal Could you add the smooth scroll as well and fix the travis build?

**Comment by lunayach** [comment]
Hi, @sanketbansal good job on the PR. Just a couple of things;

- As you have implemented smooth scrolling, a working GIF for same could be added in PR description.
- I think you could change the PR title and it's description being more specific about what you have done.

File: src/app/components/privacy-policy/privacy-policy.component.ts [hunk 3]
```diff
@@ -132,6 +132,17 @@ export class PrivacyPolicyComponent implements OnInit {
    * Scrolls to the top of the page
    */
   scrollToTop() {
-    this.document.body.scrollTop = this.document.documentElement.scrollTop = 0;
+    // this.document.body.scrollTop = this.document.documentElement.scrollTop = 0;
```
Comment (lunayach) [inline src/app/components/privacy-policy/privacy-policy.component.ts]: The commented code may be removed.

File: src/app/components/privacy-policy/privacy-policy.component.ts [hunk 4]
```diff
@@ -132,6 +132,17 @@ export class PrivacyPolicyComponent implements OnInit {
...
+    const bodyScrollRate = this.document.body.scrollTop / 100;
+    const eleScrollRate = this.document.documentElement.scrollTop / 100;
+
+    const scrollInt = setInterval(() => {
+      this.document.body.scrollTop -= bodyScrollRate;
+      this.document.documentElement.scrollTop -= eleScrollRate;
+      console.log(' Scrolling Body and Element ');
```
Comment (lunayach) [inline src/app/components/privacy-policy/privacy-policy.component.ts]: Log statement may also be removed.

**Comment by sanketbansal** [comment]
evalai_edit_0

**Comment by sanketbansal** [comment]
Added the gif and changed the title of pull request.

File: CODE_OF_CONDUCT.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# Contributor Code of Conduct
```
Comment (gautamjajoo) [inline CODE_OF_CONDUCT.md]: This has been added in previous commit! Please remove this

File: src/app/components/privacy-policy/privacy-policy.component.scss [hunk 6]
```diff
@@ -54,6 +54,7 @@
       right: 4%;
       height: 50px;
       cursor: pointer;
+      z-index: 4 !important;
```
Comment (gautamjajoo) [inline src/app/components/privacy-policy/privacy-policy.component.scss:58]: Why was this added? Decreasing the `z-index` of footer works!

**Review by gautamjajoo — COMMENTED** [review]
Everything else looks good, one minor comment.

File: src/app/components/privacy-policy/privacy-policy.component.scss (hunk 6 above)
Comment (sanketbansal) [reply src/app/components/privacy-policy/privacy-policy.component.scss:58]: yeah but just to make sure that it's z-index remains high then others elements

**Comment by Shekharrajak** [comment]
Please don't remove link https://pr-111-evalai.surge.sh/ from the PR description .

File: src/app/components/privacy-policy/privacy-policy.component.scss (hunk 6 above)
Comment (Shekharrajak) [reply src/app/components/privacy-policy/privacy-policy.component.scss:58]: Then it should be high number like 100, right?

File: src/app/components/privacy-policy/privacy-policy.component.ts (hunk 3 above)
Comment (Shekharrajak) [inline src/app/components/privacy-policy/privacy-policy.component.ts]: Unnecessary lines should be removed before committing the changes.

File: src/app/components/privacy-policy/privacy-policy.component.ts (hunk 4 above)
Comment (Shekharrajak) [inline src/app/components/privacy-policy/privacy-policy.component.ts]: `console.log` can slow down the page. It should not be committed.

File: src/app/components/privacy-policy/privacy-policy.component.ts [hunk 7]
```diff
@@ -132,6 +132,17 @@ export class PrivacyPolicyComponent implements OnInit {
...
+      if ( this.document.body.scrollTop === 0 && this.document.documentElement.scrollTop === 0 ) {
+        clearInterval(scrollInt);
```
Comment (Shekharrajak) [inline src/app/components/privacy-policy/privacy-policy.component.ts]: I am not a good fan of using pure javascript in Angular application. Isn't we can do the same using 

```
scroll(el: HTMLElement) {
    el.scrollIntoView({behavior:"smooth"});
  }

```

Also, the same method can be used to scroll to down/up when user will click on contents; written in the left sidebar. WDYT ?

**Comment by Shekharrajak** [comment]
I see a few other minor things (not related to this PR and issue) : 

1. Unnecessary x-scroll in the privacy-page
2. Scroll to top arrow should be disappeared when it is already on top.
3. When we click on left sidebar content, it should move up/down smoothly on the clicked section. 
Means when I click on "Third-party links" written in left side bar it should smoothly go to the "Third-party links" at the right.

WDYT?

**Comment by gautamjajoo** [comment]
@Shekharrajak I was about to open an issue for 3rd point but since this PR is working on almost the same purpose, @sanketbansal Could you make the changes requested by @Shekharrajak??

**Comment by sanketbansal** [comment]
ok, Will commit some changes for that issue.

File: src/app/components/privacy-policy/privacy-policy.component.ts (hunk 7 above)
Comment (sanketbansal) [reply src/app/components/privacy-policy/privacy-policy.component.ts]: have you checked this method to solve the smooth scroll issue?

**Comment by sanketbansal** [comment]
cloudcv

**Comment by sanketbansal** [comment]
fix #107

**Comment by sanketbansal** [comment]
Screenshot (44)

**Comment by Shekharrajak** [comment]
@sanketbansal , your work is good, but few points about PRs:

- All the things that describe the PR should be in PR description.  writing fix #issue_number somewhere is PR comment is bad practice. You may edit the PR description/ GIF image with the new commit.

- commit messages are very important.  fix #issue_number, having typo error is not good in the commit message, also it should be totally relevant to your changes.  You can change the commit message and/or squash the commit as per the need.

Thanks!

File: CODE_OF_CONDUCT.md [hunk 8]
```diff
@@ -1,44 +0,0 @@
-# Contributor Code of Conduct
```
Comment (gautamjajoo) [inline CODE_OF_CONDUCT.md]: Why have you deleted the Code of Conduct file? Please revert this.

Comment (sanketbansal) [reply CODE_OF_CONDUCT.md]: https://github.com/Cloud-CV/EvalAI-ngx/pull/111#discussion_r267420754
What were you asking to remove in that comment?

File: src/app/components/privacy-policy/privacy-policy.component.spec.ts [hunk 9]
```diff
@@ -37,10 +41,23 @@ describe('PrivacyPolicyComponent', () => {
...
+    de = fixture.debugElement;
     fixture.detectChanges();
   });
 
   it('should create', () => {
     expect(component).toBeTruthy();
   });
+
+  it('should have equal number of section-title elements as nav elements', () => {
+    expect(component).toBeTruthy();
```
Comment (Shekharrajak) [inline src/app/components/privacy-policy/privacy-policy.component.spec.ts]: Already present in line 49.

File: src/app/components/privacy-policy/privacy-policy.component.spec.ts [hunk 10]
```diff
@@ -37,10 +41,23 @@ describe('PrivacyPolicyComponent', () => {
...
+
+    const ALL_NAV = de.queryAll(By.css('.privacy-nav'));
+    const ALL_TARGET = de.queryAll(By.css('.privacy-section-title'));
+
+    expect(ALL_NAV.length).toBeGreaterThan(0);
+    expect(ALL_TARGET.length).toBeGreaterThan(0);
+    expect(ALL_TARGET.length).toBe(ALL_NAV.length);
```
Comment (Shekharrajak) [inline src/app/components/privacy-policy/privacy-policy.component.spec.ts:60]: This is good one.

**Review by Shekharrajak — COMMENTED** [review]
We can also compare the title of sidebar with title of right side (.privacy-section-title).

WDYT?

File: src/app/components/privacy-policy/privacy-policy.component.spec.ts (hunk 9 above)
Comment (sanketbansal) [reply src/app/components/privacy-policy/privacy-policy.component.spec.ts]: > We can also compare the title of sidebar with title of right side (.privacy-section-title).

yeah, but the thing is not every sidebar title is same as section title.

Comment (Shekharrajak) [reply src/app/components/privacy-policy/privacy-policy.component.spec.ts]: If we make it same then any issue or will it change the meaning ?

Comment (sanketbansal) [reply src/app/components/privacy-policy/privacy-policy.component.spec.ts]: No, it will not change the meaning
Should i do that?

Comment (Shekharrajak) [reply src/app/components/privacy-policy/privacy-policy.component.spec.ts]: Yes!

File: src/app/components/privacy-policy/privacy-policy.component.spec.ts [hunk 11]
```diff
@@ -37,10 +41,33 @@ describe('PrivacyPolicyComponent', () => {
...
+  });
+
+  it('should have same section title as nav element title', () => {
+
+    const ALL_NAV = de.queryAll(By.css('.privacy-nav'));
```
Comment (Shekharrajak) [inline src/app/components/privacy-policy/privacy-policy.component.spec.ts]: Line 54, 55 is sameas line 64,65. Can you apply [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) concept here?

**Review by Shekharrajak — COMMENTED** [review]
Other thanmy comment, it is looking good.

**Review by Shekharrajak — APPROVED** [review]

**Comment by Shekharrajak** [comment]
Please fix travis build error, other than that it is good to go.

**Comment by sanketbansal** [comment]
How can I see the logs of test command on travis build ?

**Comment by Shekharrajak** [comment]
Isn't it clearly mentioned here : https://travis-ci.org/Cloud-CV/EvalAI-ngx/builds/525279140#L585 ?

**Comment by sanketbansal** [comment]
yeah, I can see those logs but what I meant is the **test-log** file generated for **ng test command** .

**Comment by Shekharrajak** [comment]
@sanketbansal , why not you try to understand how travis CI build works and which command it is running for testing and what changes need to be done to check test logs?

**Comment by RishabhJain2018** [comment]
@sanketbansal Can you please look into this and try to understand the working of Travis builds. Feel free to ping @Shekharrajak @lunayach or me if you need some help.

**Comment by sanketbansal** [comment]
yeah working on the issue

**Comment by RishabhJain2018** [comment]
@sanketbansal Any updates on this?

**Comment by sanketbansal** [comment]
@RishabhJain2018  Travis build issue is resolved now.
Please review

**Comment by RishabhJain2018** [comment]
Thanks, @sanketbansal for fixing it quickly.
@Shekharrajak @lunayach Can you guys please review this?

**Review by lunayach — CHANGES_REQUESTED** [review]
@sanketbansal  Currently, the surge demo is not working. Could you rebase over the master branch? #127 got merged recently, and that should make the demo work.

^ Ignore, I updated the branch using the UI below.

**Comment by lunayach** [comment]
@RishabhJain2018 after #127 the demos should have been working. http://staging-evalai.cloudcv.org/api/ is the staging endpoint, Right?

**Comment by sanketbansal** [comment]
@lunayach  All these changes had already been done. These are outdated commits.

**Comment by RishabhJain2018** [comment]
There is a CORS issue with the staging server. I'll fix this in a day or two.

**Comment by sanketbansal** [comment]
@lunayach  are there any changes needs to done?

**Review by lunayach — COMMENTED** [review]
The code looks good. But just a small UI thing that I noticed. Shouldn't the scroll button become visible only when the user has scrolled down a bit? Currently, it just comes after some slight delay even when the user has not scrolled down.

**Comment by sanketbansal** [comment]
@lunayach I have added [these changes](https://github.com/Cloud-CV/EvalAI-ngx/pull/111#pullrequestreview-238719026). please review.

**Review by lunayach — APPROVED** [review]

**Comment by lunayach** [comment]
@RishabhJain2018 @Shekharrajak Let's get this merged!

File: .travis.yml [hunk 12]
```diff
@@ -14,14 +14,17 @@ before_install:
   - export CHROME_BIN=chromium-browser
   - export DISPLAY=:99.0
   - export DEPLOY_DOMAIN=https://pr-${TRAVIS_PULL_REQUEST}-evalai.surge.sh
-  - sh -e /etc/init.d/xvfb start
+  - /sbin/start-stop-daemon --start --quiet --pidfile /tmp/custom_xvfb_99.pid --make-pidfile --background --exec /usr/bin/Xvfb -- :99 -ac -screen 0 1280x1024x16
```
Comment (Shekharrajak) [inline .travis.yml:17]: I am not sure about this change. @sanketbansal can you please let us know what this line will do?

**Review by Shekharrajak — APPROVED** [review]
Good work @sanketbansal ! Let's wait for https://github.com/Cloud-CV/EvalAI-ngx/pull/125 then we will merge this.

File: .travis.yml (hunk 12 above)
Comment (sanketbansal) [reply .travis.yml:17]: basically, on privacy policy page we do not show **nav bar** for **small screen size** due to which nav bar is not getting **rendered** in headless chrome environment causing values coming to be **null**. So used this line to change screen size.

**Comment by Shekharrajak** [comment]
Can we merge this PR now ? @sanketbansal

**Comment by sanketbansal** [comment]
Yeah, It's working fine I think.

**Comment by Shekharrajak** [comment]
Okay! Merging it.
